tests/semantic_analyser.cpp

```diff
@@ -4035,4 +4035,69 @@ kfunc:foo { $p = path((uint8 *)0) }
   test("BEGIN { $b = buf(0, 10) }");
 }
 
+TEST(semantic_analyser, variable_declarations)
+{
+  test("BEGIN { let $a; $a = 1; }");
+  test("BEGIN { let $a: int16; $a = 1; }");
+  test("BEGIN { let $a = 1; }");
+  test("BEGIN { let $a: int16 = 1; }");
+  test("BEGIN { let $a: str_t[5] = \"hiya\"; }");
+  // If the type is specified it's strict in that future assignments
+  // need to fit into that type
+  test(R"(BEGIN { let $a: str_t[5] = "hiya"; $a = "bye"; })");
+  test(R"(BEGIN { let $a = "hiya"; $a = "longerstr"; })");
+  test("BEGIN { let $a: int16 = 1; $a = (int8)2; }");
+
+  test_error("BEGIN { let $a; }", R"(
+stdin:1:9-15: ERROR: Variable $a never assigned to.
+BEGIN { let $a; }
+        ~~~~~~
+)");
```

Comment:
It's a good question. I went back and forth on it. I don't feel strongly other than I kinda like it when I get yelled at for doing silly things. That being said, it can be annoying if you're iterating fast and run into stuff like this. Maybe I will remove it - scripting languages are supposed to be flexible. Unless anyone else has strong feelings...

Reply:
I made it a warning instead of an error.